projects/NodeBB/services/nodebb/module.nix

```diff
@@ -6,13 +6,180 @@
+            mongo = 27017;
+            postgres = 5432;
+            redis = 6379;
+          }
+          .${cfg.database.type};
+        defaultText = lib.literalMD "default port of selected database";
+        description = "Database host port.";
+      };
+
+      username = lib.mkOption {
+        type = lib.types.str;
+        default = "nodebb";
+        description = ''
+          Database user.
+
+          This is not used for Redis.
+        '';
+      };
+
+      passwordFile = lib.mkOption {
+        type = lib.types.path;
+        description = "Path to a file containing the database password.";
+      };
+
+      database = lib.mkOption {
+        type = lib.types.either lib.types.int lib.types.str;
+        default = if cfg.database.type == "redis" then 0 else "nodebb";
+        defaultText = lib.literalExpression ''if cfg.database.type == "redis" then 0 else "nodebb"'';
+      };
+
+      ssl = lib.mkOption {
+        type = lib.types.bool;
+        default = false;
+        description = ''
+          Whether to enable TLS.
+
+          This is only used for PostgreSQL.
+        '';
+      };
+    };
   };
 
   config = lib.mkIf cfg.enable {
+    systemd.services.nodebb = {
+      description = "NodeBB";
+      documentation = [ "https://docs.nodebb.org" ];
+      after = [
+        "system.slice"
+        "multi-user.target"
+      ];
+
+      environment.CONFIG = "/etc/nodebb/config.json";
+
+      serviceConfig = {
+        Type = "forking";
+        User = cfg.user;
+        Group = cfg.group;
+        StateDirectory = "nodebb";
+        ConfigurationDirectory = "nodebb";
+        WorkingDirectory = "/var/lib/nodebb";
+        PIDFile = "/var/lib/nodebb/pidfile";
+        ExecStart = "${lib.getExe pkgs.nodejs} loader.js --no-silent";
+        Restart = "always";
+      };
+
+      preStart = ''
+        ${lib.getExe pkgs.rsync} -a --chmod=u+w ${cfg.package}/lib/node_modules/nodebb/ .
+
+        # Cannot copy favicon.ico
+        mkdir -p public/uploads/system
+
+        if test ! -f $CONFIG; then
+          cp ${configFile} $CONFIG
+        else
+          # https://stackoverflow.com/a/24904276
+          ${lib.getExe pkgs.jq} -s '.[0] * .[1]' $CONFIG ${configFile} > $CONFIG.new
+          mv $CONFIG.new $CONFIG
+        fi
+
+        databasePasswordKey='."${cfg.database.type}:password"'
+        databasePassword="$(<${cfg.database.passwordFile})"
+        ${lib.getExe pkgs.jq} "$databasePasswordKey = \"$databasePassword\"" $CONFIG > $CONFIG.new
+        mv $CONFIG.new $CONFIG
+
+        export NODEBB_ADMIN_USERNAME="${cfg.admin.username}"
+        export NODEBB_ADMIN_PASSWORD="$(<${cfg.admin.passwordFile})"
+        export NODEBB_ADMIN_EMAIL="${cfg.admin.email}"
+        ./nodebb setup "$(<$CONFIG)"
+      '';
+
+      wantedBy = [ "multi-user.target" ];
+    };
+
+    users = {
+      groups.${cfg.group} = { };
+      users.${cfg.user} = {
+        inherit (cfg) group;
+        isSystemUser = true;
+      };
+    };
```

Comment:
Maybe we could run the service as a [DynamicUser](https://www.freedesktop.org/software/systemd/man/latest/systemd.exec.html#DynamicUser=).